lib/tree.rb

```diff
@@ -16,51 +16,225 @@ def initialize
+  #     end
+  #   end
+  #   return current
+  # end
+
+  # Time Complexity: O(log n)
+  # Space Complexity: O(n)
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root, key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # def find(key)
+  #   current_node = @root
+  #   if current_node.nil?
+  #     return nil
+  #   else
+  #     while true
+  #       if current_node.key == key
+  #         return current_node.value
+  #       elsif current_node.key > key
+  #         if current_node.left.nil?
+  #           return false
+  #         else
+  #           current_node = current_node.left
+  #         end
+  #       else
+  #         if current_node.right.nil?
+  #           return false
+  #         else
+  #           current_node = current_node.right
+  #         end
+  #       end
+  #     end
+  #   end
+  # end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    postorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(2n) or O(n)
   def height
-    raise NotImplementedError
+    return height_helper(@root)
   end
 
   # Optional Method
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity:
+  # Space Complexity:
   def bfs
     raise NotImplementedError
   end
 
+  def delete(key)
+    parent = nil
+    current = @root
+    # find the node, keeping track of its parent
+    result = find_with_parent_helper(current, parent, key)
+    return nil if result.nil?
+    current = result[:current]
+    parent = result[:parent]
+    # if the found node has no children
+    #   - if the node was parent.right, set parent.right to nil
+    #   - if the node was parent.left, set parent.left to nil
+    if current.left.nil? && current.right.nil?
+      (parent.right == current) \
+      ? parent.right = nil \
+      : parent.left = nil
+    # else if the found node has 1 child
+    #   - if child is .right, point parent.right to the child
+    elsif !current.right.nil? && current.left.nil?
+      (parent.right == current) \
+      ? parent.right = current.right \
+      : parent.left = current.right
+    #   - if child is .left, point parent.left to the child
+    elsif !current.left.nil? && current.right.nil?
+      (parent.right == current) \
+        ? parent.right = current.left \
+        : parent.left = current.left
+    else
+      # else if the found node has 2 children
+      #   - find the right child's leftmost leaf
+      right_child = current.right
+      leftmost_left = right_child.left
```

Comment:
I think a better variable name here is `leftmost_right = right_child`.

Note I used `right_child` instead of `right_child.left` because you don't know if the right child has a left child as well.